torch/csrc/distributed/autograd/engine/dist_engine.cpp

```diff
@@ -131,13 +139,54 @@ void DistEngine::computeDependencies(
             //  its ancestors need to be executed as well.
             if (dynamic_cast<RecvRpcBackward*>(nextFn)) {
               recvBackwardEdges.emplace_back(edge);
+            } else if (
+                auto accumulateGradFn = dynamic_cast<AccumulateGrad*>(nextFn)) {
+              TORCH_INTERNAL_ASSERT(
+                  !accumulateGradFn->variable.defined() ||
+                      nextFn ==
+                          torch::autograd::impl::try_get_grad_accumulator(
+                              accumulateGradFn->variable)
+                              .get(),
+                  "The AccumulateGrad point in the variable should be the "
+                  "same as the one in autograph");
+              // Replace an AccumulateGrad node by DistAccumulateGrad,
+              // because the former accumulates grads to the variable's '.grad'
+              // without considering the context id. That may cause data race
+              // on the '.grad', since multiple context ids may have grads for
+              // the same '.grad'.
+              auto distAccumulateGradFn = std::make_shared<DistAccumulateGrad>(
+                  // The pointer to 'AccumulateGrad' in
+                  // 'accumulateGradFn->variable' will be replaced by a pointer
+                  // to 'distAccumulateGradFn'.
+                  std::move(*accumulateGradFn),
+                  autogradContext);
+              distAccumulateGradFn->replace_grad_accumulator();
```

Comment:
Why is this needed?
Changing the `next_edge` in the graph below means that we won't even try to use the grad_accumulator.

Reply:
All output tensor/variable keeps a weak_ptr to its `AccumulateGrad`. Would it be a problem if the `AccumulateGrad` functions in autograd graph are inconsistent with the ones in tensors/variables?

Reply:
The `AccumulateGrad` Node is the one that populates the `.grad` field of a variable.
You could modify the grad_accumulator before the graph is created. But that sounds really weird as you have no way to know about it.
So what you want is just to replace the node currently in the graph to avoid accumulating in the `.grad` field.